feat: implement `forc add` and `forc remove` to add/remove dependencies

JoE11-y opened the pull request:
## Description
Close #2369 

**Summary:**
Implements functionality for `forc` to add remove dependencies through cli, similar to Cargo's `add` and `remove` commands.

**Work Done:**

* ✅ Added `forc add` command to support adding both regular and contract dependencies from path, git, IPFS, or version.
* ✅ Implemented conflict checks for version + source collisions (e.g., version + git).
* ✅ Added automatic fallback to workspace-relative path for local packages.
* ✅ Introduced `forc remove` to cleanly remove specified dependencies from the manifest.
* ✅ Modularized update logic using `DepSection` enum for unified regular/contract dependency handling.
* ✅ Used `toml_edit` to update the manifest file while preserving formatting.


## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
   - [ ] If my change requires substantial documentation changes, I have [requested support from the DevRel team](https://github.com/FuelLabs/devrel-requests/issues/new/choose)
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

JoE11-y commented:
@IGI-111 this is ready for review.

I also noticed something, i was unable to trigger the lockfile update by just updating the members_manifest, it only worked when the forc.toml was already updated with new dependencies.

zees-dev commented:
You may also want to integrate with the `forc.pub` registry - which would be a canonical endpoint to add dependencies; @sdankel, @kayagokalp

IGI-111 commented:
This looks nice but you need to add some tests and fix the CI errors before we can consider merging it.

JoE11-y commented:
Alright will do that.

On Mon, May 5, 2025, 10:47 AM IGI-111 ***@***.***> wrote:

> *IGI-111* left a comment (FuelLabs/sway#7143)
> <https://github.com/FuelLabs/sway/pull/7143#issuecomment-2850471879>
>
> This looks nice but you need to add some tests and fix the CI errors
> before we can consider merging it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FuelLabs/sway/pull/7143#issuecomment-2850471879>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANGCG5W7LKLXHHZEX2C7FW3244XRTAVCNFSM6AAAAAB4LUHOYKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNJQGQ3TCOBXHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

JoshuaBatty commented:
Hey @JoE11-y thanks for the PR. Just letting you know I plan to look at this properly in Monday.

zees-dev commented on docs/book/src/forc/commands/forc_add.md:
need docs for this

JoE11-y replied:
Alright added.

zees-dev commented on docs/book/src/forc/commands/forc_remove.md:
and need docs for this; could be minimal to start off with

JoE11-y replied:
Also added docs  here

zees-dev commented on docs/book/src/SUMMARY.md:
```diff
@@ -91,6 +92,7 @@
     - [forc parse-bytecode](./forc/commands/forc_parse-bytecode.md)
     - [forc plugins](./forc/commands/forc_plugins.md)
     - [forc predicate-root](./forc/commands/forc_predicate-root.md)
+    - [forc remove](./forc/commands/forc_remove.md)
```
I dont think this needs to be in alphabetical order - maybe others can comment here.
Since `forc remove` and `forc add` are so closely correlated; it would make more sense for these pages to be right next to each other

JoE11-y replied:
Oh alright, that makes sense.
Done.

zees-dev commented on forc-pkg/src/manifest/mod.rs:
```diff
@@ -345,8 +367,17 @@ impl DependencyDetails {
         }
+
+        if version.is_some() && path.is_some() {
+            bail!("Both version and path details provided fro same dependency");
```
```suggestion
            bail!("Both version and path details provided for same dependency");
```

zees-dev commented on forc/src/cli/shared.rs:
```diff
@@ -1,9 +1,79 @@
+        .multiple(false) // only one allowed
+        .requires("git") // require --git if any git_ref is set
+))]
+pub struct GitRef {
```
if only 1 is required; wondering if we can use `enum` here instead

JoE11-y replied:
I tried but Enums only supports unit variants.

E.g

```
pub enum Level {
    Debug,
    Info,
    Warning,
    Error,
}
```

zees-dev replied:
In this scenario; it could be worthwhile to have an intermediary struct - which enforces these constraints.
Would also make it easier to unit test.
The code below would also likely be less bug-prone as the validations for the struct -> enum conversions would be done upfront.

zees-dev commented on forc/src/cli/commands/add.rs:
```diff
@@ -0,0 +1,60 @@
+}
+
+pub(crate) fn exec(command: Command) -> ForcResult<()> {
+    let _ = command;
```
can remove this line

zees-dev commented on forc/src/cli/commands/add.rs:
```diff
@@ -0,0 +1,60 @@
+    let _ = command;
+    match forc_add::add(command) {
+        Ok(_) => Ok(()),
+        Err(e) => Err(format!("couldn't add dependencies: {}", e).as_str().into()),
```
nit:
```suggestion
        Err(e) => Err(format!("failed to add dependencies: {}", e).as_str().into()),
```

JoE11-y replied:
alright done

zees-dev commented on forc-pkg/src/manifest/manager.rs:
```diff
@@ -0,0 +1,484 @@
+}
+
+pub fn modify_dependencies(opts: ModifyOpts) -> Result<()> {
+    let cwd = if let Some(p) = opts.manifest_path.clone() {
```
wondering if we need to clone `manifest_path` - since `PathBuf::from` can take reference

JoE11-y replied:
yes this works.
resolved.

zees-dev commented on forc-pkg/src/manifest/manager.rs:
```diff
@@ -0,0 +1,484 @@
+
+    member_manifests.insert(
+        package_spec.project_name().to_string(),
+        package_spec.clone(),
```
cant see var usage after this point; could remove clone?

```suggestion
        package_spec,
```

JoE11-y replied:
Yes this works.

zees-dev commented on forc-pkg/src/manifest/manager.rs:
```diff
@@ -0,0 +1,484 @@
+        &opts.ipfs_node.clone().unwrap_or_default(),
+    );
+
+    if new_plan.is_err() {
+        std::fs::write(&package_spec_dir, backup_doc.to_string())
+            .map_err(|e| anyhow!("failed to write toml file: {}", e))?;
+        bail!(new_plan.err().unwrap());
+    }
+
+    let new_plan = new_plan?;
```
nit:

```suggestion
    let new_plan = new_plan.or_else(|e| {
        std::fs::write(&package_spec_dir, backup_doc.to_string())
            .map_err(|write_err| anyhow!("failed to write toml file: {}", write_err))?;
        Err(e)
    })?;
```

JoE11-y replied:
this is way cleaner, than what i had in mind.
resolved.

zees-dev commented on forc-pkg/src/manifest/manager.rs:
```diff
@@ -0,0 +1,484 @@
+
+    let package_spec_dir = resolve_package_path(
+        &manifest_file,
+        opts.package.clone(),
```
potentially unnecessary clone?

JoE11-y replied:
yes another clone, passed in ref and it worked.

zees-dev commented on forc-pkg/src/manifest/manager.rs:
```diff
@@ -0,0 +1,484 @@
+    let dep_spec: DepSpec = raw.parse()?;
+    let dep_name = dep_spec
+        .name
+        .clone()
```
unnecessary clone?

JoE11-y replied:
Okay, maybe it’s time I stop with the cloning 😅.

Resolved.

zees-dev commented on forc-pkg/src/manifest/manager.rs:
```diff
@@ -0,0 +1,484 @@
+            bail!("dependency spec cannot be empty");
+        }
+
+        let mut dep = DepSpec::default();
```
nit:
Move below `let Some(name) ...` block; since this var declaration is unused if the `bail` on line `267` is hit.

JoE11-y replied:
okay yh.
That makes more sense.

Resolved.

zees-dev commented on forc-pkg/src/manifest/manager.rs:
```diff
@@ -0,0 +1,484 @@
+                map.insert(name, data);
+            }
+            DepSection::Contract(map, salt_opt) => {
+                let resolved_salt = salt
+                    .or_else(|| salt_opt.clone())
+                    .map(|s| HexSalt::from_str(&s).unwrap())
+                    .unwrap_or_else(|| HexSalt(fuel_tx::Salt::default()));
```
unnecessary clone; also need to handle `unwrap`
```suggestion
                let resolved_salt = match salt.as_ref().or_else(|| salt_opt.as_ref()) {
                    Some(s) => HexSalt::from_str(s)
                        .map_err(|e| anyhow!("Invalid salt format: {}", e))?,
                    None => HexSalt(fuel_tx::Salt::default()),
                };
```

JoE11-y replied:
Okay updated function to return result.
Resolved

zees-dev commented on forc-pkg/src/manifest/manager.rs:
```diff
@@ -0,0 +1,484 @@
+                    map.remove(*dep);
+                }
+                manifest.dependencies = Some(map.clone());
+                remove_deps_from_table(doc, manifest_path, "dependencies", deps)?;
```
nit: `"dependencies"` and `"contract-dependencies"` could be set as constants

JoE11-y replied:
alright, much more cleaner.

Resolved.

zees-dev commented on forc-pkg/src/manifest/manager.rs:
```diff
@@ -0,0 +1,484 @@
+}
+
+#[cfg(test)]
+mod tests {
```
would like more unit tests for adding deps, removing deps, converting DepSpec from string, resolving deps, modify_dependencies, etc.

JoE11-y replied:
okay will work on this.

zees-dev reviewed (commented):
Thanks for PR; functionality looks good!
Highlighted some improvements; definitely would like unit tests though.

kayagokalp commented on forc/src/cli/commands/add.rs:
```diff
@@ -0,0 +1,60 @@
+
+pub(crate) fn exec(command: Command) -> ForcResult<()> {
+    let _ = command;
+    match forc_add::add(command) {
+        Ok(_) => Ok(()),
+        Err(e) => Err(format!("couldn't add dependencies: {}", e).as_str().into()),
+    }
```
I think we can do a `map_err` instead of explicit match here.

```suggestion
    forc_add::add(command).map_err(|e| format!("failed to add dependencies: {}", e))
```

JoE11-y replied:
okay looks good.
resolved

kayagokalp commented on forc/src/ops/forc_add.rs:
```diff
@@ -0,0 +1,41 @@
+
+fn opts_from_cmd(cmd: AddCommand) -> pkg::manifest::manager::ModifyOpts {
+    pkg::manifest::manager::ModifyOpts {
+        // === Action ====
```
We can remove these
```suggestion
```

JoE11-y replied:
alright done.

kayagokalp commented on forc/src/ops/forc_add.rs:
```diff
@@ -0,0 +1,41 @@
+    pkg::manifest::manager::ModifyOpts {
+        // === Action ====
+        action: Action::Add,
+        // === Manifest Options ===
```
```suggestion
```

JoE11-y replied:
alright done

kayagokalp commented on forc/src/ops/forc_add.rs:
```diff
@@ -0,0 +1,41 @@
+        // === Manifest Options ===
+        manifest_path: cmd.manifest.manisfest_path,
+
+        // === Package Selection ===
```
```suggestion
```

JoE11-y replied:
alright done

kayagokalp commented on forc/src/ops/forc_add.rs:
```diff
@@ -0,0 +1,41 @@
+        // === Package Selection ===
+        package: cmd.package.package,
+
+        // === Source ===
```
```suggestion
```

kayagokalp commented on forc/src/ops/forc_add.rs:
```diff
@@ -0,0 +1,41 @@
+        rev: cmd.source.git_ref.rev,
+        ipfs: cmd.source.ipfs,
+
+        // === Section ===
```
```suggestion
```

JoE11-y replied:
alright done

kayagokalp commented on forc/src/ops/forc_add.rs:
```diff
@@ -0,0 +1,41 @@
+        contract_deps: cmd.section.contract_deps,
+        salt: cmd.section.salt,
+
+        // === IPFS Node ===
```
```suggestion
```

JoE11-y replied:
alright done

kayagokalp commented on forc/src/ops/forc_add.rs:
```diff
@@ -0,0 +1,41 @@
+        // === IPFS Node ===
+        ipfs_node: cmd.ipfs_node,
+
+        // === Dependencies & Flags ===
```
```suggestion
```

kayagokalp commented on forc/src/ops/forc_remove.rs:
```diff
@@ -0,0 +1,41 @@
+
+fn opts_from_cmd(cmd: RemoveCommand) -> pkg::manifest::manager::ModifyOpts {
+    pkg::manifest::manager::ModifyOpts {
+        // === Action ====
```
```suggestion
```

JoE11-y replied:
alright done

kayagokalp commented on forc/src/ops/forc_remove.rs:
```diff
@@ -0,0 +1,41 @@
+    pkg::manifest::manager::ModifyOpts {
+        // === Action ====
+        action: pkg::manifest::manager::Action::Remove,
+        // === Manifest Options ===
```
```suggestion
```

JoE11-y replied:
alright done

kayagokalp commented on forc/src/ops/forc_remove.rs:
```diff
@@ -0,0 +1,41 @@
+        // === Manifest Options ===
+        manifest_path: cmd.manifest.manisfest_path,
+
+        // === Package Selection ===
```
```suggestion
```

JoE11-y replied:
alright done

kayagokalp commented on forc/src/ops/forc_remove.rs:
```diff
@@ -0,0 +1,41 @@
+        // === Package Selection ===
+        package: cmd.package.package,
+
+        // === Source ===
```
```suggestion
```

kayagokalp commented on forc/src/ops/forc_remove.rs:
```diff
@@ -0,0 +1,41 @@
+        rev: None,
+        ipfs: None,
+
+        // === Section ===
```
```suggestion
```

JoE11-y replied:
alright done.

kayagokalp commented on forc/src/ops/forc_remove.rs:
```diff
@@ -0,0 +1,41 @@
+        contract_deps: cmd.section.contract_deps,
+        salt: cmd.section.salt,
+
+        // === IPFS Node ===
```
```suggestion
```

JoE11-y replied:
alright done

kayagokalp commented on forc/src/ops/forc_remove.rs:
```diff
@@ -0,0 +1,41 @@
+        // === IPFS Node ===
+        ipfs_node: cmd.ipfs_node,
+
+        // === Dependencies & Flags ===
```
```suggestion
```

JoE11-y replied:
alright done.

kayagokalp commented on Cargo.toml:
```diff
@@ -187,6 +187,7 @@ once_cell = "1.18"
 opener = "0.7"
 parking_lot = "0.12"
 paste = "1.0"
+pathdiff = "0.2.3"
```
Is there a specific reason we need to pin to `0.2.3`? Can we keep this at `0.2`? If that is the case that is more preferable (also removing the `ALLOWED_XYZ_DEPS` entry)

JoE11-y replied:
okay i'll look into another solution, they didn't have other an X.Y only version

JoE11-y replied:
I already found another alternative that didn't require me using path diff.

kayagokalp reviewed (commented):
Thanks a lot for the PR! Looking close :) left couple of nits

JoE11-y commented:
Alright will get right to refactoring the pr 🚀.

Thanks guys

JoshuaBatty commented on forc-pkg/src/manifest/manager.rs:
Not a fan of modules called things like manager. Perhaps `dependency_modifier.rs` is a better fit?

JoE11-y replied:
haha alright will do that.

JoE11-y commented:
I've addressed all the issues raised in the review, added documentation where needed, and included relevant tests. 💪

zees-dev commented on forc-pkg/src/manifest/dep_modifier.rs:
```diff
@@ -0,0 +1,1133 @@
+    // === Package Selection ===
+    pub package: Option<String>,
+    // === Source (Add only) ===
+    pub source_path: Option<String>,
+    pub git: Option<String>,
+    pub branch: Option<String>,
+    pub tag: Option<String>,
+    pub rev: Option<String>,
+    pub ipfs: Option<String>,
```
nit: the vars can be exncapsulated in another struct; and potentially flattened

zees-dev commented on forc/src/ops/forc_add.rs:
```diff
@@ -0,0 +1,29 @@
+    Ok(())
+}
+
+fn opts_from_cmd(cmd: AddCommand) -> pkg::manifest::dep_modifier::ModifyOpts {
```
could implement the `From<AddCommand>` trait for `ModifyOpts` - in the same file.
This file could potentially be removed.

JoE11-y replied:
I've made the changes.

JoE11-y commented on forc/src/ops/forc_remove.rs:
```diff
@@ -0,0 +1,29 @@
+    Ok(())
+}
+
+fn opts_from_cmd(cmd: RemoveCommand) -> pkg::manifest::dep_modifier::ModifyOpts {
```
so something like a forc_modify ops?

zees-dev commented on forc-pkg/src/manifest/dep_modifier.rs:
```diff
@@ -0,0 +1,1133 @@
+        let git = "https://github.com/example/repo.git";
+
+        // Git + branch
+        {
```
should also add tests where git is not provided, at all.
I.e. branch only, no git; git only, no branch
Same for the other ones below.

JoE11-y replied:
Okay will add that

JoE11-y replied:
branch only or any other ref alone without git would fail.

so i'll create another test for failures.

zees-dev commented on forc-pkg/src/manifest/dep_modifier.rs:
```diff
@@ -0,0 +1,1133 @@
+    inline
+}
+
+fn remove_deps_from_table(doc: &mut DocumentMut, section: &str, deps: &[&str]) -> Result<()> {
```
nit: `section` could be an enum - with the 2 variants; the display trait impl should allow for string conversion

JoE11-y replied:
Oh alright

zees-dev commented on forc-pkg/src/manifest/dep_modifier.rs:
```diff
@@ -0,0 +1,1133 @@
+        let result = remove_deps_from_table(&mut doc, "contract-dependencies", &["foo"]);
+        assert!(result.is_err());
+        assert!(format!("{}", result.unwrap_err())
+            .contains("section [contract-dependencies] not found"));
```
hmm thinking about whether we should return the same error as if the dep is not found; rather than returning that the section does not exist; thoughts other contributors?
Wondering what cargo does.

zees-dev replied:
yea cargo returns the following (example):
"error: the dependency `<dep>` could not be found in `dependencies`"

JoE11-y replied:
Oh okay will update that.

zees-dev commented on forc-pkg/src/manifest/dep_modifier.rs:
```diff
@@ -0,0 +1,1133 @@
+
+    #[test]
+    fn test_generate_table_basic_fields() {
+        let details = DependencyDetails {
```
interesting; so if all of these are specified; what takes precedence when building

JoE11-y replied:
They won't be at runtime.
I just added it to ensure that the table populated correctly.

The checks are already handled by this validate function.

https://github.com/JoE11-y/sway/blob/feat/forc-edit-subcommands-impl/forc-pkg/src/manifest/mod.rs#L342

Which is called here

https://github.com/JoE11-y/sway/blob/feat/forc-edit-subcommands-impl/forc-pkg/src/manifest/dep_modifier.rs#L219

JoE11-y replied:
I'll add failure tests for scenario where multiple sources are added

zees-dev commented on forc/src/cli/commands/remove.rs:
```diff
@@ -0,0 +1,77 @@
+forc_util::cli_examples! {
+crate::cli::Opt {
+    [Add a dependencies => "forc remove <DEP>"]
+    [Add a contract dependency => "forc renove <DEP> --contract-dep"]
```
```suggestion
    [Add a contract dependency => "forc remove <DEP> --contract-dep"]
```

JoE11-y replied:
Okay noted.

JoE11-y commented:
**Summary of Changes**

* Resolved all issues raised in the review.
* Refactored `DepSection` to take in non mutatble reference of dependency mappings and added  support for  dynamic section names via `Display`, 
* Unified `add_deps_manifest_table` and `remove_deps_manifest_table` to write directly to the correct TOML section.
* Improved error messages for better clarity when handling dependency conflicts or missing sources.
* Added default version fallback (`"0.1.0"`) for dependencies without a specified source.
* Added comprehensive unit tests for the changes

zees-dev commented on forc-pkg/src/manifest/dep_modifier.rs:
```diff
@@ -0,0 +1,1149 @@
+    }
+}
+
+#[derive(Clone)]
+pub enum DepSection {
+    Regular,
+    Contract,
+}
+
```
nit:
```suggestion
#[derive(Clone)]
pub enum Section {
    Dependencies,
    ContractDependencies,
}
```
or
```suggestion
#[derive(Clone)]
pub enum Section {
    Deps,
    ContractDeps,
}
```

JoE11-y replied:
Alright picked Deps & ContractDeps variant

zees-dev commented on forc-pkg/src/manifest/dep_modifier.rs:
```diff
@@ -0,0 +1,1149 @@
+}
+
+impl DepSection {
+    pub fn new(contract_deps: bool) -> Self {
```
function not needed; caller can simply instantiate enum with relevant variant

JoE11-y replied:
I get the idea, I've removed the function, but I think a simple boolean flag like `--contract-dep` feels more intuitive here. Asking users to pass enum variants like `--section ContractDeps` feels a bit clunky and unnecessary, especially when we’re only toggling between two options.

With the boolean, it’s just:

```sh
forc add --contract-dep my_contract
```

That reads much better than explicitly writing out enum names. Less room for user error, and it aligns more with how most CLI tools handle this kind of toggle.


```sh
forc add --section Contract
```

zees-dev replied:
o I was under the assumption the functionality above (calling approach) remains the same - as I can see it only being referenced in 1 place:
```rs
let section = DepSection::new(opts.contract_deps);
```
where it could be replaced with:
```rs
let section = DepSection::Contract;
```

zees-dev replied:
I do agree with your points (i.e. we dont want to pass in enum variants `--section ContractDeps`) into CLI; however i dont see how remove the `new` function could be causing this.

zees-dev commented on forc-pkg/src/manifest/dep_modifier.rs:
```diff
@@ -0,0 +1,1149 @@
+/// See `forc add` help for more info.
+#[derive(Clone, Debug, Default)]
+pub struct DepSpec {
+    pub name: Option<String>,
+    pub version_req: Option<String>,
```
shouldnt these 2 fields always be required?

JoE11-y replied:
okay updated.

zees-dev commented on forc-pkg/tests/test_workspace/Forc.toml:
nit: rather than creating minimal workspace with empty files; this can dynamically be done in the tests - example here:
https://github.com/FuelLabs/sway/pull/7172/files#diff-a716c942c7d2947f1f991faad650f0f4bebab784c1f5241a32ed78bffcff3c78R1285

JoE11-y replied:
This hasn't been merged yet.

Should i just copy it?

zees-dev replied:
yeap feel free to copy; i'd imagine your code may have slight differences in workspace

JoE11-y replied:
okay will try it out.

zees-dev commented on forc-pkg/src/manifest/dep_modifier.rs:
```diff
@@ -240,74 +242,64 @@ fn resolve_dependency(
+            anyhow::bail!("Dependency spec cannot be empty");
         }
 
         let mut s = s.split('@');
```
nit: we should probably continue with trimmed output; what do you think?

JoE11-y replied:
uh strange, my current code is with trim.

zees-dev replied:
```suggestion
        let mut s = s.trim().split('@');
```

JoE11-y replied:
oh okay. done

zees-dev reviewed (approved):
Great work; thanks for this; LGTM! 🚀

kayagokalp commented on forc-pkg/src/manifest/dep_modifier.rs:
```diff
@@ -0,0 +1,1294 @@
+            details.path = Some(rel_path.to_string_lossy().to_string());
+            Dependency::Detailed(details)
+        } else {
+            // Fallback: no explicit source & not a sibling package — default to version 0.1.0.
+            // TODO: Integrate registry support (e.g., forc.pub) here.
+            Dependency::Simple("0.1.0".to_string())
```
can we instead keep a todo! here? I feel like defaulting to 0.1.0 is kind of suprising and may create very hard to debug situations

JoE11-y replied:
Okay will update to fail there

JoE11-y replied:
Done @kayagokalp

JoE11-y commented:
I removed the default semver fallback.

And just added the an error requesting user to specify version either via git or path or with version

kayagokalp commented on forc-pkg/src/manifest/dep_modifier.rs:
```diff
@@ -0,0 +1,1293 @@
+
+    let updated_package_manifest = PackageManifestFile::from_file(&package_manifest_dir)?;
+
+    member_manifests.insert(
+        updated_package_manifest.project_name().to_string(),
+        updated_package_manifest,
+    );
+
+    let new_plan = pkg::BuildPlan::from_lock_and_manifests(
+        &lock_path,
+        &member_manifests,
+        false,
+        opts.offline,
+        &opts.ipfs_node.clone().unwrap_or_default(),
+    );
```
This logic seems off to me, the dependencies are not members of build plan Members' refer to the workspace members. I will test in a bit but my guess is that this won't even work, until then i am blocking this PR so that it doesn't get merged accidentally. I'll update in a bit

Also there is probably no need to be this verbose and do these steps manually, we should be able to run `build` after modifying the manifest file, `build` will run build plan generation, lock updates, and all the extra logic you have here.

JoE11-y replied:
First off, thank you for catching that and for all the thorough reviews so far.

I really appreciate it. I wanted to share a bit of the journey behind these changes:

- I initially pulled the target package directly from the workspace’s `member_manifests()` and updated it in place, expecting that to regenerate the build plan and lockfile.

- Then I discovered that updating only the in-memory manifest wasn’t enough to trigger the lockfile update.

- So with the help of and with subsequent revisions from @zees-dev , I was able to rework the pr, to focus on updating the toml file directly, performing the checks along the way.

I’ve spent countless hours on this iteration, tweaking, testing, and making sure it behaves exactly like its sibling commands. I’m confident that this approach is rock solid. 

And yes that logic was from the old design.

simply shadowing it with the updatedPackageManifest will clear it up

` let member_manifests = updated_package_manifest.member_manifests()?;`

Thanks again for your patience and feedback. I’m excited to get this merged! ❤️

kayagokalp replied:
Hey again, what i meant was that you shouldn't need to modify `member_manifest` at all because:

1. What you are implementing here is not related to member manifests, added dependency should not be a member manifest.
2. It is not needed to do this manually, build plan generation handles all the steps you are doing manually, you just need to update the manifest file (forc.toml) and regenerate the build plan that will take care of everything :)

JoE11-y replied:
Yeah that was the old design.

I'm no longer updating the member_manifests, that was the old design.

I'm only using it to properly verify the package path.

As we know from cargo.

We can call the add command from either the root of a workspace or from the root of a package.

So if in workspace the package where the dependency is to be added/removed should be specified with the --package flag.

That insert was part of the old design where I figured I could take out the exact manifest and then update.

kayagokalp replied:
I just tested the version at 6a785495d94082fa5927ac36d245a7ccc36b21fd seems to be working, i am fine to keep it working and get it merged we can clean it later on :)

kayagokalp reviewed (approved):
Great work on this! Thanks :) Approving this as this is getting unnecessarily long PR history, and the feature seems to be working.

My final comments are:

1. The logic to add/remove packages a little too convoluted and there are areas of improvements to seriously cut down the code size.
2. `forc add` without any dep to add doesn't produce any error/warning simply does nothing silently which is not great

But given that you already put so much work on this, I think we can merge this as it is, and later on I can do a small refactor to keep this code section little bit smaller by using the existing code-paths.

zees-dev reviewed (approved):
Approving again; thanks for your contribution @JoE11-y!

Any additional work required should probably be captured in a TODO and/or GH issue - so it can be addressed at some point.

JoE11-y commented:
The forc add/remove failure without error

can be handled by adding the required field in clap for them.

```
#[clap(value_enum, value_name = "DEP_SPEC", required = true)]
pub dependencies: Vec<String>
```
    
This was a massive learning experience for me.

Thanks @zees-dev  and @kayagokalp.

JoE11-y commented:
@zees-dev @JoshuaBatty resolved the comment removal issue and the forc add without dep failure.

kayagokalp commented:
Hey @JoE11-y, can you squash your changes by any chance? It seems like our ci step is blocked, as we have a commit in this  branch (merge master commit) from a team member that recently left and the ci cla step is blocking because of that. If you can squash your commits and force push to this branch that will go away or you can override the author of each commit to this PR to be you. 

you should be able to somehting like:

```sh
git reset --soft master
git add ...
git commit -m "feat: implement forc add and forc remove to add/remove dependencies"
git push origin feat/forc-edit-subcommands-impl -f
```

JoE11-y commented:
Hi @kayagokalp all done.
